runtime/compiler/trj9/optimizer/JProfilingRecompLoopTest.cpp

```diff
@@ -0,0 +1,214 @@
+#include "infra/ILWalk.hpp"
+#include "infra/List.hpp"
+#include "optimizer/Optimization_inlines.hpp"
+#include "il/Node_inlines.hpp"
+#include "infra/Checklist.hpp"             // for TR::NodeChecklist
+#include "ras/DebugCounter.hpp"
+#include "control/Recompilation.hpp"              // for TR_Recompilation, etc
+#include "control/RecompilationInfo.hpp"              // for TR_Recompilation, etc
+#include "codegen/CodeGenerator.hpp"
+#include "optimizer/TransformUtil.hpp"            // for TransformUtil
+
+
+int32_t TR_JProfilingRecompLoopTest::recompileThreshold = 4096;
+
+
+int32_t TR_JProfilingRecompLoopTest::getLoopNestingLevel(int32_t nestingFactor)
+   {
+   if (nestingFactor == 1)
+      nestingFactor = 0;
+   
+   switch(nestingFactor)
+      {
+      case 0: return 0;
+      case 10: return 1;
+      case 100: return 2;
+      case 1000: return 3;
+      case 10000: return 4;
+      default: return 5;
+      }
+   }
+
+int32_t
+TR_JProfilingRecompLoopTest::perform()
+   {
+   if (comp()->getOption(TR_EnableJProfiling))
+      {
+      if (trace())
+         traceMsg(comp(), "JProfiling has been enabled, run JProfiling\n");
+      }
+   else if (comp()->getProfilingMode() == JProfiling)
+      {
+      if (trace())
+         traceMsg(comp(), "JProfiling has been enabled for profiling compilations, run JProfilingRecompLoopTest\n");
+      }
+   else
+      {
+      if (trace())
+         traceMsg(comp(), "JProfiling has not been enabled, skip JProfilingRecompLoopTest\n");
+      return 0;
+      }
+   RecompilationTestLocations testLocations(RecompilationTestLocationAllocator(comp()->trMemory()->currentStackRegion()));
+   TR::TreeTop *cursor = comp()->getStartTree();
+   TR::CFG *cfg = comp()->getFlowGraph();
+
+   while (cursor)
+      {
+      if (cursor->getNode()->getOpCodeValue() == TR::asynccheck && cursor->getEnclosingBlock()->getStructureOf()->getContainingLoop() != NULL)
+         {
+         int32_t loopDepth = 1;
+         TR::Block *block = cursor->getEnclosingBlock();
+         TR_RegionStructure *region = block->getParentStructureIfExists(comp()->getFlowGraph());
+         if (region)
+            region->calculateFrequencyOfExecution(&loopDepth);
+         loopDepth = getLoopNestingLevel(loopDepth);
+         testLocations.push_back(std::make_pair(cursor, loopDepth)); 
+         }
+      cursor = cursor->getNextTreeTop(); 
+      }
+   if (!testLocations.empty())
+      addRecompilationTests(comp(), testLocations);
+   return 1;
+   }
+
+
+void 
+TR_JProfilingRecompLoopTest::addRecompilationTests(TR::Compilation *comp, RecompilationTestLocations &testLocations)
+   {
+   TR_PersistentProfileInfo *profileInfo = comp->getRecompilationInfo()->findOrCreateProfileInfo();
+   TR_BlockFrequencyInfo *bfi = TR_BlockFrequencyInfo::get(profileInfo);
+   TR::CFG *cfg = comp->getFlowGraph();
+   cfg->setStructure(NULL);
+
+   for (auto testLocation = testLocations.begin(); testLocation != testLocations.end(); ++testLocation)
```

Comment:
you should should not run end() on every iteration of the loop. A more efficient style is
for (auto testLocItr = testLocations.begin(), testLocEnd = testLocations.end(); testLocItr != testLocEnd; ++testLocItr)